src/ComputeSharp.D2D1Interop/Shaders/Translation/D2D1ShaderCompiler.cs

```diff
@@ -0,0 +1,255 @@
+using TerraFX.Interop.Windows;
+
+#pragma warning disable IDE1006
+
+namespace ComputeSharp.D2D1Interop.Shaders.Translation;
+
+/// <summary>
+/// A <see langword="class"/> that uses the FXC APIs to compile D2D1 pixel shaders.
+/// </summary>
+internal static unsafe class D2D1ShaderCompiler
+{
+    /// <summary>
+    /// Compiles a new HLSL shader from the input source code.
+    /// </summary>
+    /// <param name="source">The HLSL source code to compile.</param>
+    /// <param name="enableLinkingSupport">Whether to enable linking support for the shader.</param>
+    /// <returns>The bytecode for the compiled shader.</returns>
+    public static ComPtr<ID3DBlob> Compile(ReadOnlySpan<char> source, bool enableLinkingSupport)
+    {
+        // Compile the standalone D2D1 full shader
+        using ComPtr<ID3DBlob> d3DBlobFullShader = CompileD2DFullShader(source);
+
+        if (!enableLinkingSupport)
+        {
+            return d3DBlobFullShader.Move();
+        }
+
+        // Compile the export function and embed it as private data if requested
+        using ComPtr<ID3DBlob> d3DBlobFunction = CompileD2DFunction(source);
+        using ComPtr<ID3DBlob> d3DBlobLinked = EmbedD2DFunctionPrivateData(d3DBlobFullShader.Get(), d3DBlobFunction.Get());
+
+        return d3DBlobLinked.Move();
+    }
+
+    /// <summary>
+    /// Compiles a D2D1 pixel shader with <c>D2D_FULL_SHADER</c>.
+    /// </summary>
+    ///<param name="source">The HLSL source code to compile.</param>
+    /// <returns>The bytecode for the compiled shader.</returns>
+    private static ComPtr<ID3DBlob> CompileD2DFullShader(ReadOnlySpan<char> source)
+    {
+        // Encode the HLSL source to ASCII
+        int maxLength = Encoding.ASCII.GetMaxByteCount(source.Length);
+        byte[] buffer = ArrayPool<byte>.Shared.Rent(maxLength);
+        int writtenBytes = Encoding.ASCII.GetBytes(source, buffer);
+
+        using ComPtr<ID3DBlob> d3DBlobBytecode = default;
+        using ComPtr<ID3DBlob> d3DBlobErrors = default;
+
+        int hResult;
+
+        fixed (byte* bufferPtr = buffer)
+        {
+            // Prepare the macros for full shader compilation:
+            //
+            // -D D2D_FULL_SHADER
+            // -D D2D_ENTRY=Execute
+            D3D_SHADER_MACRO* macros = stackalloc D3D_SHADER_MACRO[]
+            {
+                new()
+                {
+                    Name = (sbyte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(ASCII.D2D_FULL_SHADER)),
+                    Definition = (sbyte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(ASCII.NullTerminator))
+                },
+                new()
+                {
+                    Name = (sbyte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(ASCII.D2D_ENTRY)),
+                    Definition = (sbyte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(ASCII.Execute))
+                },
+                new()
+            };
+
+            // Compile the shader with -ps_5_0 -O3 -We
```

Comment:
I'd also like a way to specify other shader models, including `ps_4_0_level_9_1`